IllegalStateException: Chunk not found in v1.4.200/9

After some work I ends with IllegalStateException: Chunk not found. Database v1.4.200/9 is working in tcp server mode, java 15, Linux version 4.19.0-10-amd64. Database file is small, cca 20MB with some blobs, only a few tables with a few rows.

Any suggestions to avoid this?

Best regards, 
Ondrej.

```
Exception in thread "main" org.h2.jdbc.JdbcSQLNonTransientException: Obecná chyba: "java.lang.IllegalStateException: Chunk 783 not found [1.4.200/9]"
General error: "java.lang.IllegalStateException: Chunk 783 not found [1.4.200/9]" [50000-200]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:505)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:429)
	at org.h2.message.DbException.get(DbException.java:194)
	at org.h2.message.DbException.convert(DbException.java:347)
	at org.h2.engine.Database.openDatabase(Database.java:333)
	at org.h2.engine.Database.<init>(Database.java:301)
	at org.h2.engine.Engine.openSession(Engine.java:74)
	at org.h2.engine.Engine.openSession(Engine.java:192)
	at org.h2.engine.Engine.createSessionAndValidate(Engine.java:171)
	at org.h2.engine.Engine.createSession(Engine.java:166)
	at org.h2.server.TcpServerThread.run(TcpServerThread.java:168)
	at java.base/java.lang.Thread.run(Thread.java:835)
Caused by: java.lang.IllegalStateException: Chunk 783 not found [1.4.200/9]
	at org.h2.mvstore.DataUtils.newIllegalStateException(DataUtils.java:950)
	at org.h2.mvstore.MVStore.getChunk(MVStore.java:1230)
	at org.h2.mvstore.MVStore.readBufferForPage(MVStore.java:1214)
	at org.h2.mvstore.MVStore.readPage(MVStore.java:2209)
	at org.h2.mvstore.MVMap.readPage(MVMap.java:672)
	at org.h2.mvstore.MVMap.readOrCreateRootPage(MVMap.java:688)
	at org.h2.mvstore.MVMap.setRootPos(MVMap.java:682)
	at org.h2.mvstore.MVStore.openMap(MVStore.java:576)
	at org.h2.mvstore.MVStore.openMap(MVStore.java:535)
	at org.h2.mvstore.MVStore.openMap(MVStore.java:516)
	at org.h2.mvstore.db.LobStorageMap.init(LobStorageMap.java:87)
	at org.h2.engine.Database.open(Database.java:791)
	at org.h2.engine.Database.openDatabase(Database.java:307)
	... 7 more
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:505)
	at org.h2.engine.SessionRemote.done(SessionRemote.java:611)
	at org.h2.engine.SessionRemote.initTransfer(SessionRemote.java:147)
	at org.h2.engine.SessionRemote.connectServer(SessionRemote.java:435)
	at org.h2.engine.SessionRemote.connectEmbeddedOrServer(SessionRemote.java:321)
	at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:173)
	at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:152)
	at org.h2.Driver.connect(Driver.java:69)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:677)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:228)
	at org.h2.tools.Shell.runTool(Shell.java:148)
	at org.h2.tools.Shell.main(Shell.java:81)
```

Exception in trace file: 

```
2020-09-29 03:48:46 database: closed
2020-09-29 03:48:46 database: disconnected session #18 261
2020-09-29 03:48:46 command: slow query: 2 631 ms
2020-09-29 11:48:13 database: flush
org.h2.message.DbException: Obecná chyba: "java.lang.IllegalStateException: java.lang.NullPointerException: Cannot invoke ""org.h2.mvstore.Chunk.accountForRemovedPage(int, boolean, long, long)"" because ""<local8>"" is null [1.4.200/3]"
General error: "java.lang.IllegalStateException: java.lang.NullPointerException: Cannot invoke ""org.h2.mvstore.Chunk.accountForRemovedPage(int, boolean, long, long)"" because ""<local8>"" is null [1.4.200/3]" [50000-200]
	at org.h2.message.DbException.get(DbException.java:194)
	at org.h2.message.DbException.convert(DbException.java:347)
	at org.h2.mvstore.db.MVTableEngine$1.uncaughtException(MVTableEngine.java:93)
	at org.h2.mvstore.MVStore.handleException(MVStore.java:2877)
	at org.h2.mvstore.MVStore.panic(MVStore.java:481)
	at org.h2.mvstore.MVStore.store(MVStore.java:1338)
	at org.h2.mvstore.MVStore.store(MVStore.java:1312)
	at org.h2.mvstore.MVStore.tryCommit(MVStore.java:1272)
	at org.h2.mvstore.MVStore.writeInBackground(MVStore.java:2780)
	at org.h2.mvstore.MVStore$BackgroundWriterThread.run(MVStore.java:3290)
Caused by: org.h2.jdbc.JdbcSQLNonTransientException: Obecná chyba: "java.lang.IllegalStateException: java.lang.NullPointerException: Cannot invoke ""org.h2.mvstore.Chunk.accountForRemovedPage(int, boolean, long, long)"" because ""<local8>"" is null [1.4.200/3]"
General error: "java.lang.IllegalStateException: java.lang.NullPointerException: Cannot invoke ""org.h2.mvstore.Chunk.accountForRemovedPage(int, boolean, long, long)"" because ""<local8>"" is null [1.4.200/3]" [50000-200]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:505)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:429)
	... 10 more
Caused by: java.lang.IllegalStateException: java.lang.NullPointerException: Cannot invoke "org.h2.mvstore.Chunk.accountForRemovedPage(int, boolean, long, long)" because "<local8>" is null [1.4.200/3]
	at org.h2.mvstore.DataUtils.newIllegalStateException(DataUtils.java:950)
	... 5 more
Caused by: java.lang.NullPointerException: Cannot invoke "org.h2.mvstore.Chunk.accountForRemovedPage(int, boolean, long, long)" because "<local8>" is null
	at org.h2.mvstore.MVStore.acceptChunkOccupancyChanges(MVStore.java:1612)
	at org.h2.mvstore.MVStore.storeNow(MVStore.java:1433)
	at org.h2.mvstore.MVStore.store(MVStore.java:1334)
	... 4 more
2020-09-29 12:43:27 database: opening /home/tomcat/.typografikon-application-products/typografikon-application-products (build 200)

```

The same problem occurs with trace log:

```
2020-09-29 03:48:43 lock: 1 exclusive write lock unlock SYS 
2020-09-29 03:48:43 lock: 1 exclusive write lock requesting for SYS 
2020-09-29 03:48:43 lock: 1 exclusive write lock added for SYS 
2020-09-29 03:48:43 lock: 1 exclusive write lock unlock SYS 
2020-09-29 03:48:46 lock: 1 shared read lock unlock SYS 
2020-09-29 03:48:46 lock: 2 shared read lock unlock SYS 
2020-09-29 03:48:46 database: closed
2020-09-29 03:48:46 database: disconnected session #18 261
2020-09-29 03:48:46 command: slow query: 2 631 ms
2020-09-29 11:48:13 database: flush
org.h2.message.DbException: Obecná chyba: "java.lang.IllegalStateException: java.lang.NullPointerException: Cannot invoke ""org.h2.mvstore.Chunk.accountForRemovedPage(int, boolean, long, long)"" because ""<local8>"" is null [1.4.200/3]"
General error: "java.lang.IllegalStateException: java.lang.NullPointerException: Cannot invoke ""org.h2.mvstore.Chunk.accountForRemovedPage(int, boolean, long, long)"" because ""<local8>"" is null [1.4.200/3]" [50000-200]
	at org.h2.message.DbException.get(DbException.java:194)
	at org.h2.message.DbException.convert(DbException.java:347)
	at org.h2.mvstore.db.MVTableEngine$1.uncaughtException(MVTableEngine.java:93)
	at org.h2.mvstore.MVStore.handleException(MVStore.java:2877)
	at org.h2.mvstore.MVStore.panic(MVStore.java:481)
	at org.h2.mvstore.MVStore.store(MVStore.java:1338)
	at org.h2.mvstore.MVStore.store(MVStore.java:1312)
	at org.h2.mvstore.MVStore.tryCommit(MVStore.java:1272)
	at org.h2.mvstore.MVStore.writeInBackground(MVStore.java:2780)
	at org.h2.mvstore.MVStore$BackgroundWriterThread.run(MVStore.java:3290)
Caused by: org.h2.jdbc.JdbcSQLNonTransientException: Obecná chyba: "java.lang.IllegalStateException: java.lang.NullPointerException: Cannot invoke ""org.h2.mvstore.Chunk.accountForRemovedPage(int, boolean, long, long)"" because ""<local8>"" is null [1.4.200/3]"
General error: "java.lang.IllegalStateException: java.lang.NullPointerException: Cannot invoke ""org.h2.mvstore.Chunk.accountForRemovedPage(int, boolean, long, long)"" because ""<local8>"" is null [1.4.200/3]" [50000-200]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:505)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:429)
	... 10 more
Caused by: java.lang.IllegalStateException: java.lang.NullPointerException: Cannot invoke "org.h2.mvstore.Chunk.accountForRemovedPage(int, boolean, long, long)" because "<local8>" is null [1.4.200/3]
	at org.h2.mvstore.DataUtils.newIllegalStateException(DataUtils.java:950)
	... 5 more
Caused by: java.lang.NullPointerException: Cannot invoke "org.h2.mvstore.Chunk.accountForRemovedPage(int, boolean, long, long)" because "<local8>" is null
	at org.h2.mvstore.MVStore.acceptChunkOccupancyChanges(MVStore.java:1612)
	at org.h2.mvstore.MVStore.storeNow(MVStore.java:1433)
	at org.h2.mvstore.MVStore.store(MVStore.java:1334)
	... 4 more
2020-09-29 12:43:27 database: opening /home/tomcat/.typografikon-application-products/typografikon-application-products (build 200)
2020-09-29 12:43:27 lock: 1 exclusive write lock requesting for SYS 
2020-09-29 12:43:27 lock: 1 exclusive write lock added for SYS 
2020-09-29 12:43:27 lock: 1 exclusive write lock unlock SYS 
2020-09-29 12:43:27 lock: 1 exclusive write lock requesting for SYS 
2020-09-29 12:43:27 lock: 1 exclusive write lock added for SYS 
2020-09-29 12:43:27 lock: 1 exclusive write lock unlock SYS 

```

Thanks for replies. Something is wrong, now I got error below. After every failure I restore sql dump to freshly created database. Sql backup I created previously with command SCRIPT DROP TO dump.sql COMPRESSION GZIP CHARSET 'UTF-8'.

Should I use latest version from master branch? Should I update also client side jar or only server side? 

O.

```
javax.servlet.ServletException: org.jdbi.v3.core.ConnectionException: org.h2.jdbc.JdbcSQLNonTransientException: Obecná chyba: "java.lang.IllegalStateException: Negative position -3907; p=172073569759041, c=chunk:272,block:7fffffffffffffff,len:7fffffff,liveMax:ffffffffffff9a80,livePages:fffffea4,map:1a7,max:26a0,next:1183,pages:13,root:9c8000027489,time:9cc9b0,unused:bc69b0,unusedAtVersion:2e6,version:272,pinCount:0 [1.4.200/6]"
General error: "java.lang.IllegalStateException: Negative position -3907; p=172073569759041, c=chunk:272,block:7fffffffffffffff,len:7fffffff,liveMax:ffffffffffff9a80,livePages:fffffea4,map:1a7,max:26a0,next:1183,pages:13,root:9c8000027489,time:9cc9b0,unused:bc69b0,unusedAtVersion:2e6,version:272,pinCount:0 [1.4.200/6]" [50000-200]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:505)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:429)
	at org.h2.message.DbException.get(DbException.java:194)
	at org.h2.message.DbException.convert(DbException.java:347)
	at org.h2.engine.Database.openDatabase(Database.java:333)
	at org.h2.engine.Database.<init>(Database.java:301)
	at org.h2.engine.Engine.openSession(Engine.java:74)
	at org.h2.engine.Engine.openSession(Engine.java:192)
	at org.h2.engine.Engine.createSessionAndValidate(Engine.java:171)
	at org.h2.engine.Engine.createSession(Engine.java:166)
	at org.h2.server.TcpServerThread.run(TcpServerThread.java:168)
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.lang.IllegalStateException: Negative position -3907; p=172073569759041, c=chunk:272,block:7fffffffffffffff,len:7fffffff,liveMax:ffffffffffff9a80,livePages:fffffea4,map:1a7,max:26a0,next:1183,pages:13,root:9c8000027489,time:9cc9b0,unused:bc69b0,unusedAtVersion:2e6,version:272,pinCount:0 [1.4.200/6]
	at org.h2.mvstore.DataUtils.newIllegalStateException(DataUtils.java:950)
	at org.h2.mvstore.Chunk.readBufferForPage(Chunk.java:345)
	at org.h2.mvstore.MVStore.readBufferForPage(MVStore.java:1214)
	at org.h2.mvstore.MVStore.readPage(MVStore.java:2209)
	at org.h2.mvstore.MVMap.readPage(MVMap.java:672)
	at org.h2.mvstore.MVMap.readOrCreateRootPage(MVMap.java:688)
	at org.h2.mvstore.MVMap.setRootPos(MVMap.java:682)
	at org.h2.mvstore.MVStore.openMap(MVStore.java:576)
	at org.h2.mvstore.MVStore.openMap(MVStore.java:535)
	at org.h2.mvstore.MVStore.openMap(MVStore.java:516)
	at org.h2.mvstore.db.LobStorageMap.init(LobStorageMap.java:87)
	at org.h2.engine.Database.open(Database.java:791)
	at org.h2.engine.Database.openDatabase(Database.java:307)
	... 7 more

	app.filter.jsp.DaoFilter.doFilter(DaoFilter.java:130)
	javax.servlet.http.HttpFilter.doFilter(HttpFilter.java:52)
```

--- a/h2/src/main/org/h2/engine/Database.java
+++ b/h2/src/main/org/h2/engine/Database.java
@@ -416,7 +416,6 @@ public final class Database implements DataHandler, CastDataProvider {
                 }
             }
             lobStorage = dbSettings.mvStore ? new LobStorageMap(this) : new LobStorageBackend(this);
-            lobStorage.init();
             systemSession.commit(true);
             trace.info("opened {0}", databaseName);
             if (persistent) {

--- a/h2/src/main/org/h2/mvstore/db/LobStorageMap.java
+++ b/h2/src/main/org/h2/mvstore/db/LobStorageMap.java
@@ -32,7 +32,6 @@ import org.h2.value.Value;
 import org.h2.value.ValueLob;
 import org.h2.value.ValueLobDatabase;
 import org.h2.value.ValueLobInMemory;
-import org.jetbrains.annotations.NotNull;
 
 /**
  * This class stores LOB objects in the database, in maps. This is the back-end

--- a/h2/src/main/org/h2/mvstore/db/LobStorageMap.java
+++ b/h2/src/main/org/h2/mvstore/db/LobStorageMap.java
@@ -13,7 +13,6 @@ import java.nio.charset.StandardCharsets;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Map.Entry;
-import java.util.concurrent.atomic.AtomicBoolean;
 
 import org.h2.api.ErrorCode;
 import org.h2.engine.Database;
@@ -43,9 +42,6 @@ public final class LobStorageMap implements LobStorageInterface
 
     private final Database database;
     private final MVStore mvStore;
-
-    private final AtomicBoolean init = new AtomicBoolean();
-
     private final Object nextLobIdSync = new Object();
     private long nextLobId;
 
@@ -57,7 +53,7 @@ public final class LobStorageMap implements LobStorageInterface
      * Value: { streamStoreId (byte[]), tableId (int),
      * byteCount (long), hash (long) }.
      */
-    private MVMap<Long, Object[]> lobMap;
+    private final MVMap<Long, Object[]> lobMap;
 
     /**
      * The reference map. It is used to remove data from the stream store: if no
@@ -67,9 +63,9 @@ public final class LobStorageMap implements LobStorageInterface
      * Key: { streamStoreId (byte[]), lobId (long) }.
      * Value: true (boolean).
      */
-    private MVMap<Object[], Boolean> refMap;
+    private final MVMap<Object[], Boolean> refMap;
 
-    private StreamStore streamStore;
+    private final StreamStore streamStore;
 
 
     public LobStorageMap(Database database) {
@@ -81,34 +77,20 @@ public final class LobStorageMap implements LobStorageInterface
         } else {
             mvStore = s.getMvStore();
         }
-    }
-
-    @Override
-    public void init() {
-        deregisterVersionUsage(initialize());
-    }
-
-    private MVStore.TxCounter initialize() {
         MVStore.TxCounter txCounter = mvStore.registerVersionUsage();
-        if (init.compareAndSet(false, true)) {
-            try {
-                lobMap = mvStore.openMap("lobMap");
-                refMap = mvStore.openMap("lobRef");
+        try {
+            lobMap = mvStore.openMap("lobMap");
+            refMap = mvStore.openMap("lobRef");
 
-                /* The stream store data map.
-                 *
-                 * Key: stream store block id (long).
-                 * Value: data (byte[]).
-                 */
-                MVMap<Long, byte[]> dataMap = mvStore.openMap("lobData");
-                streamStore = new StreamStore(dataMap);
-                // garbage collection of the last blocks
-                if (database.isReadOnly()) {
-                    return txCounter;
-                }
-                if (dataMap.isEmpty()) {
-                    return txCounter;
-                }
+            /* The stream store data map.
+             *
+             * Key: stream store block id (long).
+             * Value: data (byte[]).
+             */
+            MVMap<Long, byte[]> dataMap = mvStore.openMap("lobData");
+            streamStore = new StreamStore(dataMap);
+            // garbage collection of the last blocks
+            if (!database.isReadOnly() && !dataMap.isEmpty()) {
                 // search for the last block
                 // (in theory, only the latest lob can have unreferenced blocks,
                 // but the latest lob could be a copy of another one, and
@@ -146,21 +128,15 @@ public final class LobStorageMap implements LobStorageInterface
                 if (last != null) {
                     streamStore.setNextKey(last + 1);
                 }
-            } catch (Throwable ex) {
-                mvStore.deregisterVersionUsage(txCounter);
-                throw ex;
             }
+        } finally {
+            mvStore.deregisterVersionUsage(txCounter);
         }
-        return txCounter;
-    }
-
-    private void deregisterVersionUsage(MVStore.TxCounter txCounter) {
-        mvStore.deregisterVersionUsage(txCounter);
     }
 
     @Override
     public ValueLob createBlob(InputStream in, long maxLength) {
-        MVStore.TxCounter txCounter = initialize();
+        MVStore.TxCounter txCounter = mvStore.registerVersionUsage();
         int type = Value.BLOB;
         try {
             if (maxLength != -1
@@ -185,13 +161,13 @@ public final class LobStorageMap implements LobStorageInterface
         } catch (IOException e) {
             throw DbException.convertIOException(e, null);
         } finally {
-            deregisterVersionUsage(txCounter);
+            mvStore.deregisterVersionUsage(txCounter);
         }
     }
 
     @Override
     public ValueLob createClob(Reader reader, long maxLength) {
-        MVStore.TxCounter txCounter = initialize();
+        MVStore.TxCounter txCounter = mvStore.registerVersionUsage();
         int type = Value.CLOB;
         try {
             // we multiple by 3 here to get the worst-case size in bytes
@@ -227,7 +203,7 @@ public final class LobStorageMap implements LobStorageInterface
         } catch (IOException e) {
             throw DbException.convertIOException(e, null);
         } finally {
-            deregisterVersionUsage(txCounter);
+            mvStore.deregisterVersionUsage(txCounter);
         }
     }
 
@@ -270,7 +246,7 @@ public final class LobStorageMap implements LobStorageInterface
 
     @Override
     public ValueLob copyLob(ValueLob old_, int tableId, long length) {
-        MVStore.TxCounter txCounter = initialize();
+        MVStore.TxCounter txCounter = mvStore.registerVersionUsage();
         try {
             ValueLobDatabase old = (ValueLobDatabase) old_;
             int type = old.getValueType();
@@ -295,14 +271,14 @@ public final class LobStorageMap implements LobStorageInterface
             }
             return lob;
         } finally {
-            deregisterVersionUsage(txCounter);
+            mvStore.deregisterVersionUsage(txCounter);
         }
     }
 
     @Override
     public InputStream getInputStream(long lobId, long byteCount)
             throws IOException {
-        MVStore.TxCounter txCounter = initialize();
+        MVStore.TxCounter txCounter = mvStore.registerVersionUsage();
         try {
             Object[] value = lobMap.get(lobId);
             if (value == null) {
@@ -313,26 +289,26 @@ public final class LobStorageMap implements LobStorageInterface
             return new FilterInputStream(inputStream) {
                 @Override
                 public int read(byte[] b, int off, int len) throws IOException {
-                    MVStore.TxCounter txCounter = initialize();
+                    MVStore.TxCounter txCounter = mvStore.registerVersionUsage();
                     try {
                         return super.read(b, off, len);
                     } finally {
-                        deregisterVersionUsage(txCounter);
+                        mvStore.deregisterVersionUsage(txCounter);
                     }
                 }
 
                 @Override
                 public int read() throws IOException {
-                    MVStore.TxCounter txCounter = initialize();
+                    MVStore.TxCounter txCounter = mvStore.registerVersionUsage();
                     try {
                         return super.read();
                     } finally {
-                        deregisterVersionUsage(txCounter);
+                        mvStore.deregisterVersionUsage(txCounter);
                     }
                 }
             };
         } finally {
-            deregisterVersionUsage(txCounter);
+            mvStore.deregisterVersionUsage(txCounter);
         }
     }
 
@@ -341,7 +317,7 @@ public final class LobStorageMap implements LobStorageInterface
         if (mvStore.isClosed()) {
             return;
         }
-        MVStore.TxCounter txCounter = initialize();
+        MVStore.TxCounter txCounter = mvStore.registerVersionUsage();
         try {
             // this might not be very efficient -
             // to speed it up, we would need yet another map
@@ -361,20 +337,20 @@ public final class LobStorageMap implements LobStorageInterface
                 removeAllForTable(LobStorageFrontend.TABLE_RESULT);
             }
         } finally {
-            deregisterVersionUsage(txCounter);
+            mvStore.deregisterVersionUsage(txCounter);
         }
     }
 
     @Override
     public void removeLob(ValueLob lob_) {
-        MVStore.TxCounter txCounter = initialize();
+        MVStore.TxCounter txCounter = mvStore.registerVersionUsage();
         try {
             ValueLobDatabase lob = (ValueLobDatabase) lob_;
             int tableId = lob.getTableId();
             long lobId = lob.getLobId();
             removeLob(tableId, lobId);
         } finally {
-            deregisterVersionUsage(txCounter);
+            mvStore.deregisterVersionUsage(txCounter);
         }
     }
 

--- a/h2/src/main/org/h2/mvstore/db/LobStorageMap.java
+++ b/h2/src/main/org/h2/mvstore/db/LobStorageMap.java
@@ -320,6 +320,16 @@ public final class LobStorageMap implements LobStorageInterface
                         deregisterVersionUsage(txCounter);
                     }
                 }
+
+                @Override
+                public int read() throws IOException {
+                    MVStore.TxCounter txCounter = initialize();
+                    try {
+                        return super.read();
+                    } finally {
+                        deregisterVersionUsage(txCounter);
+                    }
+                }
             };
         } finally {
             deregisterVersionUsage(txCounter);

--- a/h2/src/main/org/h2/mvstore/db/LobStorageMap.java
+++ b/h2/src/main/org/h2/mvstore/db/LobStorageMap.java
@@ -5,6 +5,7 @@
  */
 package org.h2.mvstore.db;
 
+import java.io.FilterInputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.Reader;
@@ -31,6 +32,7 @@ import org.h2.value.Value;
 import org.h2.value.ValueLob;
 import org.h2.value.ValueLobDatabase;
 import org.h2.value.ValueLobInMemory;
+import org.jetbrains.annotations.NotNull;
 
 /**
  * This class stores LOB objects in the database, in maps. This is the back-end
@@ -308,7 +310,18 @@ public final class LobStorageMap implements LobStorageInterface
                 throw DbException.get(ErrorCode.LOB_CLOSED_ON_TIMEOUT_1, "" + lobId);
             }
             byte[] streamStoreId = (byte[]) value[0];
-            return streamStore.get(streamStoreId);
+            InputStream inputStream = streamStore.get(streamStoreId);
+            return new FilterInputStream(inputStream) {
+                @Override
+                public int read(@NotNull byte[] b, int off, int len) throws IOException {
+                    MVStore.TxCounter txCounter = initialize();
+                    try {
+                        return super.read(b, off, len);
+                    } finally {
+                        deregisterVersionUsage(txCounter);
+                    }
+                }
+            };
         } finally {
             deregisterVersionUsage(txCounter);
         }

--- a/h2/src/main/org/h2/mvstore/db/LobStorageMap.java
+++ b/h2/src/main/org/h2/mvstore/db/LobStorageMap.java
@@ -313,7 +313,7 @@ public final class LobStorageMap implements LobStorageInterface
             InputStream inputStream = streamStore.get(streamStoreId);
             return new FilterInputStream(inputStream) {
                 @Override
-                public int read(@NotNull byte[] b, int off, int len) throws IOException {
+                public int read(byte[] b, int off, int len) throws IOException {
                     MVStore.TxCounter txCounter = initialize();
                     try {
                         return super.read(b, off, len);

--- a/h2/src/main/org/h2/mvstore/db/LobStorageMap.java
+++ b/h2/src/main/org/h2/mvstore/db/LobStorageMap.java
@@ -12,6 +12,7 @@ import java.nio.charset.StandardCharsets;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Map.Entry;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 import org.h2.api.ErrorCode;
 import org.h2.engine.Database;
@@ -35,13 +36,14 @@ import org.h2.value.ValueLobInMemory;
  * This class stores LOB objects in the database, in maps. This is the back-end
  * i.e. the server side of the LOB storage.
  */
-public class LobStorageMap implements LobStorageInterface {
-
+public final class LobStorageMap implements LobStorageInterface
+{
     private static final boolean TRACE = false;
 
     private final Database database;
+    private final MVStore mvStore;
 
-    private boolean init;
+    private final AtomicBoolean init = new AtomicBoolean();
 
     private final Object nextLobIdSync = new Object();
     private long nextLobId;
@@ -68,83 +70,96 @@ public class LobStorageMap implements LobStorageInterface {
 
     private StreamStore streamStore;
 
+
     public LobStorageMap(Database database) {
         this.database = database;
-    }
-
-    @Override
-    public void init() {
-        if (init) {
-            return;
-        }
-        init = true;
         Store s = database.getStore();
-        MVStore mvStore;
         if (s == null) {
             // in-memory database
             mvStore = MVStore.open(null);
         } else {
             mvStore = s.getMvStore();
         }
-        lobMap = mvStore.openMap("lobMap");
-        refMap = mvStore.openMap("lobRef");
+    }
 
-        /* The stream store data map.
-         *
-         * Key: stream store block id (long).
-         * Value: data (byte[]).
-         */
-        MVMap<Long, byte[]> dataMap = mvStore.openMap("lobData");
-        streamStore = new StreamStore(dataMap);
-        // garbage collection of the last blocks
-        if (database.isReadOnly()) {
-            return;
-        }
-        if (dataMap.isEmpty()) {
-            return;
-        }
-        // search for the last block
-        // (in theory, only the latest lob can have unreferenced blocks,
-        // but the latest lob could be a copy of another one, and
-        // we don't know that, so we iterate over all lobs)
-        long lastUsedKey = -1;
-        for (Entry<Long, Object[]> e : lobMap.entrySet()) {
-            long lobId = e.getKey();
-            Object[] v = e.getValue();
-            byte[] id = (byte[]) v[0];
-            long max = streamStore.getMaxBlockKey(id);
-            // a lob may not have a referenced blocks if data is kept inline
-            if (max != -1 && max > lastUsedKey) {
-                lastUsedKey = max;
-                if (TRACE) {
-                    trace("lob " + lobId + " lastUsedKey=" + lastUsedKey);
+    @Override
+    public void init() {
+        deregisterVersionUsage(initialize());
+    }
+
+    private MVStore.TxCounter initialize() {
+        MVStore.TxCounter txCounter = mvStore.registerVersionUsage();
+        if (init.compareAndSet(false, true)) {
+            try {
+                lobMap = mvStore.openMap("lobMap");
+                refMap = mvStore.openMap("lobRef");
+
+                /* The stream store data map.
+                 *
+                 * Key: stream store block id (long).
+                 * Value: data (byte[]).
+                 */
+                MVMap<Long, byte[]> dataMap = mvStore.openMap("lobData");
+                streamStore = new StreamStore(dataMap);
+                // garbage collection of the last blocks
+                if (database.isReadOnly()) {
+                    return txCounter;
                 }
+                if (dataMap.isEmpty()) {
+                    return txCounter;
+                }
+                // search for the last block
+                // (in theory, only the latest lob can have unreferenced blocks,
+                // but the latest lob could be a copy of another one, and
+                // we don't know that, so we iterate over all lobs)
+                long lastUsedKey = -1;
+                for (Entry<Long, Object[]> e : lobMap.entrySet()) {
+                    long lobId = e.getKey();
+                    Object[] v = e.getValue();
+                    byte[] id = (byte[]) v[0];
+                    long max = streamStore.getMaxBlockKey(id);
+                    // a lob may not have a referenced blocks if data is kept inline
+                    if (max != -1 && max > lastUsedKey) {
+                        lastUsedKey = max;
+                        if (TRACE) {
+                            trace("lob " + lobId + " lastUsedKey=" + lastUsedKey);
+                        }
+                    }
+                }
+                if (TRACE) {
+                    trace("lastUsedKey=" + lastUsedKey);
+                }
+                // delete all blocks that are newer
+                while (true) {
+                    Long last = dataMap.lastKey();
+                    if (last == null || last <= lastUsedKey) {
+                        break;
+                    }
+                    if (TRACE) {
+                        trace("gc " + last);
+                    }
+                    dataMap.remove(last);
+                }
+                // don't re-use block ids, except at the very end
+                Long last = dataMap.lastKey();
+                if (last != null) {
+                    streamStore.setNextKey(last + 1);
+                }
+            } catch (Throwable ex) {
+                mvStore.deregisterVersionUsage(txCounter);
+                throw ex;
             }
         }
-        if (TRACE) {
-            trace("lastUsedKey=" + lastUsedKey);
-        }
-        // delete all blocks that are newer
-        while (true) {
-            Long last = dataMap.lastKey();
-            if (last == null || last <= lastUsedKey) {
-                break;
-            }
-            if (TRACE) {
-                trace("gc " + last);
-            }
-            dataMap.remove(last);
-        }
-        // don't re-use block ids, except at the very end
-        Long last = dataMap.lastKey();
-        if (last != null) {
-            streamStore.setNextKey(last + 1);
-        }
+        return txCounter;
+    }
+
+    private void deregisterVersionUsage(MVStore.TxCounter txCounter) {
+        mvStore.deregisterVersionUsage(txCounter);
     }
 
     @Override
     public ValueLob createBlob(InputStream in, long maxLength) {
-        init();
+        MVStore.TxCounter txCounter = initialize();
         int type = Value.BLOB;
         try {
             if (maxLength != -1
@@ -168,12 +183,14 @@ public class LobStorageMap implements LobStorageInterface {
             throw DbException.get(ErrorCode.OBJECT_CLOSED, e);
         } catch (IOException e) {
             throw DbException.convertIOException(e, null);
+        } finally {
+            deregisterVersionUsage(txCounter);
         }
     }
 
     @Override
     public ValueLob createClob(Reader reader, long maxLength) {
-        init();
+        MVStore.TxCounter txCounter = initialize();
         int type = Value.CLOB;
         try {
             // we multiple by 3 here to get the worst-case size in bytes
@@ -208,6 +225,8 @@ public class LobStorageMap implements LobStorageInterface {
             throw DbException.get(ErrorCode.OBJECT_CLOSED, e);
         } catch (IOException e) {
             throw DbException.convertIOException(e, null);
+        } finally {
+            deregisterVersionUsage(txCounter);
         }
     }
 
@@ -250,75 +269,91 @@ public class LobStorageMap implements LobStorageInterface {
 
     @Override
     public ValueLob copyLob(ValueLob old_, int tableId, long length) {
-        init();
-        ValueLobDatabase old = (ValueLobDatabase) old_;
-        int type = old.getValueType();
-        long oldLobId = old.getLobId();
-        long oldLength = old.getType().getPrecision();
-        if (oldLength != length) {
-            throw DbException.getInternalError("Length is different");
+        MVStore.TxCounter txCounter = initialize();
+        try {
+            ValueLobDatabase old = (ValueLobDatabase) old_;
+            int type = old.getValueType();
+            long oldLobId = old.getLobId();
+            long oldLength = old.getType().getPrecision();
+            if (oldLength != length) {
+                throw DbException.getInternalError("Length is different");
+            }
+            Object[] value = lobMap.get(oldLobId);
+            value = value.clone();
+            byte[] streamStoreId = (byte[]) value[0];
+            long lobId = generateLobId();
+            value[1] = tableId;
+            lobMap.put(lobId, value);
+            Object[] key = {streamStoreId, lobId};
+            refMap.put(key, Boolean.TRUE);
+            ValueLob lob = ValueLobDatabase.create(
+                    type, database, tableId, lobId, length);
+            if (TRACE) {
+                trace("copy " + old.getTableId() + "/" + old.getLobId() +
+                        " > " + tableId + "/" + lobId);
+            }
+            return lob;
+        } finally {
+            deregisterVersionUsage(txCounter);
         }
-        Object[] value = lobMap.get(oldLobId);
-        value = value.clone();
-        byte[] streamStoreId = (byte[]) value[0];
-        long lobId = generateLobId();
-        value[1] = tableId;
-        lobMap.put(lobId, value);
-        Object[] key = { streamStoreId, lobId };
-        refMap.put(key, Boolean.TRUE);
-        ValueLob lob = ValueLobDatabase.create(
-                type, database, tableId, lobId, length);
-        if (TRACE) {
-            trace("copy " + old.getTableId() + "/" + old.getLobId() +
-                    " > " + tableId + "/" + lobId);
-        }
-        return lob;
     }
 
     @Override
     public InputStream getInputStream(long lobId, long byteCount)
             throws IOException {
-        init();
-        Object[] value = lobMap.get(lobId);
-        if (value == null) {
-            throw DbException.get(ErrorCode.LOB_CLOSED_ON_TIMEOUT_1, "" + lobId);
+        MVStore.TxCounter txCounter = initialize();
+        try {
+            Object[] value = lobMap.get(lobId);
+            if (value == null) {
+                throw DbException.get(ErrorCode.LOB_CLOSED_ON_TIMEOUT_1, "" + lobId);
+            }
+            byte[] streamStoreId = (byte[]) value[0];
+            return streamStore.get(streamStoreId);
+        } finally {
+            deregisterVersionUsage(txCounter);
         }
-        byte[] streamStoreId = (byte[]) value[0];
-        return streamStore.get(streamStoreId);
     }
 
     @Override
     public void removeAllForTable(int tableId) {
-        init();
-        if (database.getStore().getMvStore().isClosed()) {
+        if (mvStore.isClosed()) {
             return;
         }
-        // this might not be very efficient -
-        // to speed it up, we would need yet another map
-        ArrayList<Long> list = new ArrayList<>();
-        for (Entry<Long, Object[]> e : lobMap.entrySet()) {
-            Object[] value = e.getValue();
-            int t = (Integer) value[1];
-            if (t == tableId) {
-                list.add(e.getKey());
+        MVStore.TxCounter txCounter = initialize();
+        try {
+            // this might not be very efficient -
+            // to speed it up, we would need yet another map
+            ArrayList<Long> list = new ArrayList<>();
+            for (Entry<Long, Object[]> e : lobMap.entrySet()) {
+                Object[] value = e.getValue();
+                int t = (Integer) value[1];
+                if (t == tableId) {
+                    list.add(e.getKey());
+                }
             }
-        }
-        for (long lobId : list) {
-            removeLob(tableId, lobId);
-        }
-        if (tableId == LobStorageFrontend.TABLE_ID_SESSION_VARIABLE) {
-            removeAllForTable(LobStorageFrontend.TABLE_TEMP);
-            removeAllForTable(LobStorageFrontend.TABLE_RESULT);
+            for (long lobId : list) {
+                removeLob(tableId, lobId);
+            }
+            if (tableId == LobStorageFrontend.TABLE_ID_SESSION_VARIABLE) {
+                removeAllForTable(LobStorageFrontend.TABLE_TEMP);
+                removeAllForTable(LobStorageFrontend.TABLE_RESULT);
+            }
+        } finally {
+            deregisterVersionUsage(txCounter);
         }
     }
 
     @Override
     public void removeLob(ValueLob lob_) {
-        init();
-        ValueLobDatabase lob = (ValueLobDatabase) lob_;
-        int tableId = lob.getTableId();
-        long lobId = lob.getLobId();
-        removeLob(tableId, lobId);
+        MVStore.TxCounter txCounter = initialize();
+        try {
+            ValueLobDatabase lob = (ValueLobDatabase) lob_;
+            int tableId = lob.getTableId();
+            long lobId = lob.getLobId();
+            removeLob(tableId, lobId);
+        } finally {
+            deregisterVersionUsage(txCounter);
+        }
     }
 
     private void removeLob(int tableId, long lobId) {

--- a/h2/src/main/org/h2/pagestore/db/LobStorageBackend.java
+++ b/h2/src/main/org/h2/pagestore/db/LobStorageBackend.java
@@ -102,98 +102,82 @@ public class LobStorageBackend implements LobStorageInterface {
     private final CompressTool compress = CompressTool.getInstance();
     private long[] hashBlocks;
 
-    private boolean init;
 
     public LobStorageBackend(Database database) {
         this.database = database;
-    }
-
-    @Override
-    public void init() {
-        if (init) {
-            return;
-        }
-        synchronized (database) {
-            // have to check this again or we might miss an update on another
-            // thread
-            if (init) {
-                return;
-            }
-            init = true;
-            String systemUserName = database.getSystemUser().getName();
-            SessionLocal session = database.getLobSession();
-            session.getTrace().setLevel(TraceSystem.OFF);
-            conn = new JdbcConnection(session, systemUserName, Constants.CONN_URL_INTERNAL);
-            session = database.getSystemSession();
-            session.getTrace().setLevel(TraceSystem.OFF);
-            JdbcConnection initConn = new JdbcConnection(session, systemUserName, Constants.CONN_URL_INTERNAL);
-            try {
-                Statement stat = initConn.createStatement();
-                // stat.execute("SET UNDO_LOG 0");
-                // stat.execute("SET REDO_LOG_BINARY 0");
-                boolean create = true, update = false;
-                PreparedStatement prep = initConn.prepareStatement(
-                        "SELECT ZERO() FROM INFORMATION_SCHEMA.COLUMNS WHERE " +
-                        "TABLE_SCHEMA=? AND TABLE_NAME=? AND COLUMN_NAME=?");
+        String systemUserName = database.getSystemUser().getName();
+        SessionLocal session = database.getLobSession();
+        session.getTrace().setLevel(TraceSystem.OFF);
+        conn = new JdbcConnection(session, systemUserName, Constants.CONN_URL_INTERNAL);
+        session = database.getSystemSession();
+        session.getTrace().setLevel(TraceSystem.OFF);
+        JdbcConnection initConn = new JdbcConnection(session, systemUserName, Constants.CONN_URL_INTERNAL);
+        try {
+            Statement stat = initConn.createStatement();
+            // stat.execute("SET UNDO_LOG 0");
+            // stat.execute("SET REDO_LOG_BINARY 0");
+            boolean create = true, update = false;
+            PreparedStatement prep = initConn.prepareStatement(
+                    "SELECT ZERO() FROM INFORMATION_SCHEMA.COLUMNS WHERE " +
+                    "TABLE_SCHEMA=? AND TABLE_NAME=? AND COLUMN_NAME=?");
+            prep.setString(1, "INFORMATION_SCHEMA");
+            prep.setString(2, "LOB_MAP");
+            prep.setString(3, "POS");
+            ResultSet rs;
+            rs = prep.executeQuery();
+            if (rs.next()) {
+                prep = initConn.prepareStatement(
+                        "SELECT ZERO() FROM INFORMATION_SCHEMA.TABLES WHERE " +
+                        "TABLE_SCHEMA=? AND TABLE_NAME=?");
                 prep.setString(1, "INFORMATION_SCHEMA");
-                prep.setString(2, "LOB_MAP");
-                prep.setString(3, "POS");
-                ResultSet rs;
+                prep.setString(2, "LOB_DATA");
                 rs = prep.executeQuery();
                 if (rs.next()) {
-                    prep = initConn.prepareStatement(
-                            "SELECT ZERO() FROM INFORMATION_SCHEMA.TABLES WHERE " +
-                            "TABLE_SCHEMA=? AND TABLE_NAME=?");
+                    create = false;
+                    prep = initConn.prepareStatement("SELECT DATA_TYPE FROM INFORMATION_SCHEMA.COLUMNS WHERE "
+                            + "TABLE_SCHEMA=? AND TABLE_NAME=? AND COLUMN_NAME=?");
                     prep.setString(1, "INFORMATION_SCHEMA");
                     prep.setString(2, "LOB_DATA");
+                    prep.setString(3, "DATA");
                     rs = prep.executeQuery();
-                    if (rs.next()) {
-                        create = false;
-                        prep = initConn.prepareStatement("SELECT DATA_TYPE FROM INFORMATION_SCHEMA.COLUMNS WHERE "
-                                + "TABLE_SCHEMA=? AND TABLE_NAME=? AND COLUMN_NAME=?");
-                        prep.setString(1, "INFORMATION_SCHEMA");
-                        prep.setString(2, "LOB_DATA");
-                        prep.setString(3, "DATA");
-                        rs = prep.executeQuery();
-                        rs.next();
-                        if ("BINARY".equals(rs.getString(1))) {
-                            update = true;
-                        }
+                    rs.next();
+                    if ("BINARY".equals(rs.getString(1))) {
+                        update = true;
                     }
                 }
-                if (create) {
-                    stat.execute("CREATE CACHED TABLE IF NOT EXISTS " + LOBS +
-                            "(ID BIGINT PRIMARY KEY, BYTE_COUNT BIGINT, `TABLE` INT) HIDDEN");
-                    stat.execute("CREATE INDEX IF NOT EXISTS " +
-                            "INFORMATION_SCHEMA.INDEX_LOB_TABLE ON " +
-                            LOBS + "(`TABLE`)");
-                    stat.execute("CREATE CACHED TABLE IF NOT EXISTS " + LOB_MAP +
-                            "(LOB BIGINT, SEQ INT, POS BIGINT, HASH INT, " +
-                            "BLOCK BIGINT, PRIMARY KEY(LOB, SEQ)) HIDDEN");
-                    stat.execute("ALTER TABLE " + LOB_MAP +
-                            " RENAME TO " + LOB_MAP + " HIDDEN");
-                    stat.execute("ALTER TABLE " + LOB_MAP +
-                            " ADD IF NOT EXISTS POS BIGINT BEFORE HASH");
-                    // TODO the column name OFFSET was used in version 1.3.156,
-                    // so this can be remove in a later version
-                    stat.execute("ALTER TABLE " + LOB_MAP +
-                            " DROP COLUMN IF EXISTS \"OFFSET\"");
-                    stat.execute("CREATE INDEX IF NOT EXISTS " +
-                            "INFORMATION_SCHEMA.INDEX_LOB_MAP_DATA_LOB ON " +
-                            LOB_MAP + "(BLOCK, LOB)");
-                    stat.execute("CREATE CACHED TABLE IF NOT EXISTS " +
-                            LOB_DATA +
-                            "(BLOCK BIGINT PRIMARY KEY, COMPRESSED INT, DATA VARBINARY) HIDDEN");
-                } else if (update) {
-                    stat.execute("ALTER TABLE " + LOB_DATA + " ALTER COLUMN DATA SET DATA TYPE VARBINARY");
-                }
-                rs = stat.executeQuery("SELECT MAX(BLOCK) FROM " + LOB_DATA);
-                rs.next();
-                nextBlock = rs.getLong(1) + 1;
-                stat.close();
-            } catch (SQLException e) {
-                throw DbException.convert(e);
             }
+            if (create) {
+                stat.execute("CREATE CACHED TABLE IF NOT EXISTS " + LOBS +
+                        "(ID BIGINT PRIMARY KEY, BYTE_COUNT BIGINT, `TABLE` INT) HIDDEN");
+                stat.execute("CREATE INDEX IF NOT EXISTS " +
+                        "INFORMATION_SCHEMA.INDEX_LOB_TABLE ON " +
+                        LOBS + "(`TABLE`)");
+                stat.execute("CREATE CACHED TABLE IF NOT EXISTS " + LOB_MAP +
+                        "(LOB BIGINT, SEQ INT, POS BIGINT, HASH INT, " +
+                        "BLOCK BIGINT, PRIMARY KEY(LOB, SEQ)) HIDDEN");
+                stat.execute("ALTER TABLE " + LOB_MAP +
+                        " RENAME TO " + LOB_MAP + " HIDDEN");
+                stat.execute("ALTER TABLE " + LOB_MAP +
+                        " ADD IF NOT EXISTS POS BIGINT BEFORE HASH");
+                // TODO the column name OFFSET was used in version 1.3.156,
+                // so this can be remove in a later version
+                stat.execute("ALTER TABLE " + LOB_MAP +
+                        " DROP COLUMN IF EXISTS \"OFFSET\"");
+                stat.execute("CREATE INDEX IF NOT EXISTS " +
+                        "INFORMATION_SCHEMA.INDEX_LOB_MAP_DATA_LOB ON " +
+                        LOB_MAP + "(BLOCK, LOB)");
+                stat.execute("CREATE CACHED TABLE IF NOT EXISTS " +
+                        LOB_DATA +
+                        "(BLOCK BIGINT PRIMARY KEY, COMPRESSED INT, DATA VARBINARY) HIDDEN");
+            } else if (update) {
+                stat.execute("ALTER TABLE " + LOB_DATA + " ALTER COLUMN DATA SET DATA TYPE VARBINARY");
+            }
+            rs = stat.executeQuery("SELECT MAX(BLOCK) FROM " + LOB_DATA);
+            rs.next();
+            nextBlock = rs.getLong(1) + 1;
+            stat.close();
+        } catch (SQLException e) {
+            throw DbException.convert(e);
         }
     }
 
@@ -215,7 +199,6 @@ public class LobStorageBackend implements LobStorageInterface {
 
     @Override
     public void removeAllForTable(int tableId) {
-        init();
         try {
             String sql = "SELECT ID FROM " + LOBS + " WHERE `TABLE` = ?";
             PreparedStatement prep = prepare(sql);
@@ -346,7 +329,6 @@ public class LobStorageBackend implements LobStorageInterface {
     @Override
     public InputStream getInputStream(long lobId, long byteCount) throws IOException {
         try {
-            init();
             assertNotHolds(conn.getSession());
             // see locking discussion at the top
             synchronized (database) {
@@ -466,8 +448,7 @@ public class LobStorageBackend implements LobStorageInterface {
         synchronized (database) {
             synchronized (conn.getSession()) {
                 try {
-                    init();
-                    ValueLob v = null;
+                    ValueLob v;
                     if (!old.isRecoveryReference()) {
                         long lobId = getNextLobId();
                         String sql = "INSERT INTO " + LOB_MAP +
@@ -593,13 +574,11 @@ public class LobStorageBackend implements LobStorageInterface {
 
     @Override
     public ValueLob createBlob(InputStream in, long maxLength) {
-        init();
         return addLob(in, maxLength, Value.BLOB, null);
     }
 
     @Override
     public ValueLob createClob(Reader reader, long maxLength) {
-        init();
         long max = maxLength == -1 ? Long.MAX_VALUE : maxLength;
         CountingReaderInputStream in = new CountingReaderInputStream(reader, max);
         return addLob(in, Long.MAX_VALUE, Value.CLOB, in);

--- a/h2/src/main/org/h2/store/LobStorageFrontend.java
+++ b/h2/src/main/org/h2/store/LobStorageFrontend.java
@@ -88,10 +88,4 @@ public class LobStorageFrontend implements LobStorageInterface {
         // to read a block while writing something)
         return ValueLobFile.createTempClob(reader, maxLength, sessionRemote);
     }
-
-    @Override
-    public void init() {
-        // nothing to do
-    }
-
 }

--- a/h2/src/main/org/h2/store/LobStorageInterface.java
+++ b/h2/src/main/org/h2/store/LobStorageInterface.java
@@ -68,15 +68,9 @@ public interface LobStorageInterface {
     void removeAllForTable(int tableId);
 
     /**
-     * Initialize the lob storage.
-     */
-    void init();
-
-    /**
      * Whether the storage is read-only
      *
      * @return true if yes
      */
     boolean isReadOnly();
-
 }

--- a/h2/src/test/org/h2/test/unit/TestValueMemory.java
+++ b/h2/src/test/org/h2/test/unit/TestValueMemory.java
@@ -418,12 +418,5 @@ public class TestValueMemory extends TestBase implements DataHandler {
             // to read a block while writing something)
             return ValueLobFile.createTempClob(reader, maxLength, TestValueMemory.this);
         }
-
-        @Override
-        public void init() {
-            // nothing to do
-        }
-
     }
-
 }

